### ISSUE TITLE
raft: server may try sending a snapshot that it just dropped

### ISSUE THREAD

**Issue by kbr-** [scylladb/scylla#9407]
In `server_impl::io_fiber`, after obtaining a snapshot descriptor from `fsm`, `io_fiber` will ask `_state_machine` to drop the previous snapshot:
```
                // Drop previous snapshot since it is no longer used
                logger.trace("[{}] io_fiber dropping snapshot {}", _id, old_id);
                 _state_machine->drop_snapshot(old_id);
```
(`logger.trace` call added by me).

However, in the same batch there may be a message generated by `fsm` with this snapshot ID, and `io_fiber` will try to send it:
```
            for (auto&& m : batch.messages) {
                try {
                    send_message(m.first, std::move(m.second));
```

```
template <typename Message>
void server_impl::send_message(server_id id, Message m) {
    std::visit([this, id] (auto&& m) {
        using T = std::decay_t<decltype(m)>;
```
...
```
        } else if constexpr (std::is_same_v<T, install_snapshot>) {
            _stats.install_snapshot_sent++;
            // Send in the background.
            send_snapshot(id, std::move(m));
```
...
```
void server_impl::send_snapshot(server_id dst, install_snapshot&& snp) {
    seastar::abort_source as;
    uint64_t id = _next_snapshot_transfer_id++;
    future<> f = _rpc->send_snapshot(dst, std::move(snp), as).then_wrapped([this, dst, id] (future<snapshot_reply> f) {
```

An implementation of `rpc` will probably assume that the snapshot is still available in the state machine. Perhaps this is an incorrect assumption and `rpc` should handle missing snapshots (e.g. by returning an error immediately), in which case the implementation of `rpc::send_snapshot` in test/raft/randomized_nemesis.cc is incorrect. In this case the interface comment should probably mention this possibility and the implementation should be fixed. Otherwise we have a bug in the server code.

The race can be reproduced relatively easily with this branch:
https://github.com/kbr-/scylla/commits/drop-snapshot-race
suggested seed to run randomized_nemesis_test with: `666612166` (will vastly increase the probability of hitting the race).

Attaching an example trace.
[drop-snapshot-race.txt](https://github.com/scylladb/scylla/files/7252204/drop-snapshot-race.txt)

cc @kostja @gleb-cloudius

**Comment by kbr-** [comment]
Most likely introduced in f9f859ac40b6a96be67b8cb21447f8810aa2522a. Before that change, the snapshot would be dropped in `applier_fiber`, which could only happen after sending messages (since stuff is pushed to applier_fiber after sending messages).

**Comment by gleb-cloudius** [comment]
A snapshot can be removed before or even during sending. The bug is in the test. Documentation should mention that indeed.

**Comment by kbr-** [comment]
> Documentation should mention that indeed.

Then perhaps we should wait with closing this issue until documentation is fixed?

**Comment by kostja** [comment]
Gleb, how do you suggest the test is fixed exactly? I keep trying to document the snapshot semantics and keep learning new things, including with this issue. How can the state machine account for its snapshots if not via the commands to drop snapshot issued by Raft? If Raft instructs the state machine to drop its snapshot, how can it then instruct it to send it? Please provide clarification here how you envision this is supposed to work.

**Comment by gleb-cloudius** [comment]
> > Documentation should mention that indeed.
> 
> Then perhaps we should wait with closing this issue until documentation is fixed?

OK. lest rename it as documentation issue then.

**Comment by gleb-cloudius** [comment]
> Gleb, how do you suggest the test is fixed exactly? I keep trying to document the snapshot semantics and keep learning new things, including with this issue. How can the state machine account for its snapshots if not via the commands to drop snapshot issued by Raft? If Raft instructs the state machine to drop its snapshot, how can it then instruct it to send it? Please provide clarification here how you envision this is supposed to work.

As Kamil implemented it in its test. If a snapshot id that needs to be sent is not present throw an error from rpc::send_snasphot

**Comment by gleb-cloudius** [comment]
> > > Documentation should mention that indeed.
> > 
> > 
> > Then perhaps we should wait with closing this issue until documentation is fixed?
> 
> OK. lest rename it as documentation issue then.

Looks like only you can do it.

**Comment by kostja** [comment]
@gleb-cloudius why not simply move taking the snapshot after sending the messages? There is no requirement that this happens before.

**Comment by gleb-cloudius** [comment]
Taking snapshot? It's in the different fiber. Also a snapshot may be replaced by an incoming message. May be you mean dropping a snapshot. How will it help? Sending a snapshot is an async operation, so you will remove a snapshot just after starting its transfer (and this assumes that rpc::send_snapshot uses snapshot_id before any preemption point otherwise you will have exactly same problem, and this is too much to assume from an implementation). And why even start a transfer of an outdated snapshot?

**Comment by kostja** [comment]
Yes, I mean updating the snapshot descriptor and dropping the snapshot. At least the server won't feed a dead snapshot descriptor into rpc. Pre-emption may happen, but it won't drop the descriptor. If preemption happens inside rpc, it's up to rpc to handle it - obviously there is no guarantee that the snapshot descriptor is still there if it happens. But at least we give the implementation a chance to reference the object before it's dropped.

Your second argument, that this snapshot is outdates, is more to the point. If we begin sending this snapshot, and drop it right away, then we also are likely to truncate the log, and then we'll need another snapshot.

Then perhaps the best thing is to check if message's descriptor is valid in send() and drop the message right away if it's no longer latest? That would spare implementations from having to check it.

**Comment by gleb-cloudius** [comment]
On Sun, Oct 03, 2021 at 05:07:05AM -0700, Konstantin Osipov wrote:
> Yes, I mean updating the snapshot descriptor and dropping the snapshot.
Updating the snapshot descriptor have to happen before sending otherwise
you may transfer a snapshot that was not saved. 

> At least the server won't feed a dead snapshot descriptor into rpc.
If we define how it should be handled I do not see any problem with
that.

> Pre-emption may happen, but it won't drop the descriptor.
Raft will drop it.

> If preemption happens inside rpc, it's up to rpc to handle it - obviously there is no guarantee that the snapshot descriptor is still there if it happens. But at least we give the implementation a chance to reference the object before it's dropped.
Why would it want to reference it in the first place? If snapshot
disappears it should not be sent.

> 
> Your second argument, that this snapshot is outdates, is more to the point. If we begin sending this snapshot, and drop it right away, then we also are likely to truncate the log, and then we'll need another snapshot.
> 
Exactly. And returning an error will make sure that another one will be
send instead.

> Then perhaps the best thing is to check if message's descriptor is valid in send() and drop the message right away if it's no longer latest? That would spare implementations from having to check it.
> 
Implementation will have to check it anyway. It will have to loop it up
somewhere and do something if it is not found. If a caller guaranties
that it may never happen then the implementation may assert() otherwise
it may throw, so I do not see the reason to check it twice.

--
			Gleb.

**Comment by kostja** [comment]
@kbr- I think it's best if you update our internal docs (raft/*.md) to reflect the issue. Could you take a moment to do it please and we'll close this one?
